PIVX-Project/PIVX
# [TierTwo] Signing Request (DIP 7)

This PR enables the possibility of performing Quorums Signing Sessions:
Each time a signing session is started a quorum is deterministically chosen and any given masternode will take part to the session (i.e. sign the and relay the session message) only if it is also a member of the chosen quorum.

Network side threshold recovered signatures are sent using the inventory system with the new message type MSG_QUORUM_RECOVERED_SIG;

Signature shares instead are batched and sent directly without the inventory system, the corresponding message types are: QSIGSHARESINV QGETSIGSHARES QBSIGSHARES QSIGREC.

For testers first of all:

- Create a local regtest network
- Send SPORK21 and mine enough blocks to activate v6.0
- Create 3 deterministic masternodes
- Wait for a quorum to be successfully created and mined on chain
- You are finally ready to test this feature! 

Any masternode can try to sign (for example the following rpc command can be used)
signsession 100 "0000000000000000000000000000000000000000000000000000000000000001" "0000000000000000000000000000000000000000000000000000000000000002"

The first argument is the quorum type (100 is the value for the test quorum)
The second argument is the signing session id
The third argument is the message signed

For any given session id a masternode can sign at most 1 message.
If at least 2 of the 3 members of the quorum sign the same session id with the same message a threshold signature is generated. To check that the threshold signature has been generated the following command can be used:

hasrecoverysignature 100 "0000000000000000000000000000000000000000000000000000000000000001" "0000000000000000000000000000000000000000000000000000000000000002"

(As in the previous case 00....1 is a session id 00...2 is the message signed)

I have personally spent days testing this feature + wrote a functional test that simulates different scenarios
